python/cairo-addons/src/cairo_addons/testing/hooks.py

```diff
@@ -0,0 +1,228 @@
+    deselect_by_keyword(items, config)
+    deselect_by_mark(items, config)
+
+    # Collect only is used by the IDE to collect tests, at this point
+    # we don't want to compile the cairo files
+    if config.option.collectonly:
+        yield
+        return
+
+    tests_to_skip = config.cache.get(f"cairo_run/{CACHED_TESTS_FILE}", [])
+    session.cairo_files = {}
+    session.cairo_programs = {}
+    session.main_paths = {}
+    session.test_hashes = {}
+    cairo_items = [
+        item
+        for item in items
+        if (
+            hasattr(item, "fixturenames")
+            and set(item.fixturenames)
+            & {
+                "cairo_file",
+                "main_path",
+                "cairo_program",
+                "cairo_run",
+            }
+        )
+    ]
+
+    # Distribute compilation using modulo
+    worker_count = getattr(config, "workerinput", {}).get("workercount", 1)
+    worker_id = getattr(config, "workerinput", {}).get("workerid", "master")
+    worker_index = int(worker_id[2:]) if worker_id != "master" else 0
+    fspaths = sorted(list({item.fspath for item in cairo_items}))
+    for fspath in fspaths[worker_index::worker_count]:
+        session.cairo_files[fspath] = get_cairo_file(fspath)
+        session.main_paths[fspath] = get_main_path(session.cairo_files[fspath])
+        dump_path = get_dump_path(session, fspath)
+        session.cairo_programs[fspath] = get_cairo_program(
+            session.cairo_files[fspath],
+            session.main_paths[fspath],
+            dump_path,
+        )
+
+    # Wait for all workers to finish
+    missing = set(fspaths) - set(fspaths[worker_index::worker_count])
+    while missing:
+        logger.info(f"Waiting for {len(missing)} compilations artifacts to be ready")
+        missing_new = set()
+        for fspath in missing:
+            if fspath not in session.cairo_files:
+                session.cairo_files[fspath] = get_cairo_file(fspath)
+            if fspath not in session.main_paths:
+                session.main_paths[fspath] = get_main_path(session.cairo_files[fspath])
+            if fspath not in session.cairo_programs:
+                dump_path = get_dump_path(session, fspath)
+                if dump_path.exists():
+                    session.cairo_programs[fspath] = get_cairo_program(
+                        session.cairo_files[fspath],
+                        session.main_paths[fspath],
+                        dump_path,
+                    )
+                else:
+                    missing_new.add(fspath)
+        missing = missing_new
+        time.sleep(0.25)
+
+    # Select tests
+    for item in cairo_items:
+        cairo_program = session.cairo_programs[item.fspath]
+        test_hash = xxhash.xxh64(
+            program_hash(cairo_program)
+            + file_hash(item.fspath)
+            + item.nodeid.encode()
+            + file_hash(Path(__file__).parent / "runner.py")
+        ).hexdigest()
```

Comment:
see before we had also 

```
            + file_hash(Path(__file__).parent / "utils" / "serde.py")
            + file_hash(Path(__file__).parent / "utils" / "args_gen.py")
            + file_hash(Path(__file__).parent / "utils" / "strategies.py")
```

but I didn't know how to keep this in the context of this refacto